extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -65,6 +66,22 @@ public HttpAuthenticator(Instance<HttpAuthenticationMechanism> instance,
         if (mechanisms.isEmpty()) {
             this.mechanisms = new HttpAuthenticationMechanism[] { new NoAuthenticationMechanism() };
         } else {
+            final String basicMech = "Basic";
+            Collections.sort(mechanisms, new Comparator<HttpAuthenticationMechanism>() {
+
+                @Override
+                public int compare(HttpAuthenticationMechanism mech1, HttpAuthenticationMechanism mech2) {
+                    HttpCredentialTransport credTransport1 = mech1.getCredentialTransport();
+                    if (credTransport1 != null && basicMech.equalsIgnoreCase(credTransport1.getTypeTarget())) {
+                        return -1;
+                    }
+                    HttpCredentialTransport credTransport2 = mech2.getCredentialTransport();
+                    if (credTransport2 != null && basicMech.equalsIgnoreCase(credTransport2.getTypeTarget())) {
```

Comment:
I think we should have a priority method on HttpAuthenticationMechanism instead, to make this more flexible.

Reply:
@stuartwdouglas sure, that can work as well; so `Basic` would get a priority `2`, all other mechanisms shipped with Quarkus - `1` (as the default value for the custom ones to pick it up) - and then we'll sort in the descending order for Basic to be there first ?
Note, when some hints are available (ex `Authorization: Bearer`) then it will be a bearer mechanism which will be selected for producing a challenge; similarly if a path specific mechanism is `bearer` then it will also be selected.

So the only problem this PR is trying to address is to get a Basic challenge when no credentials are provided.

I'm fine with looking at a `priority` idea to addressing this problem - and perhaps we can follow up with more improvements related to `priority`

Reply:
@stuartwdouglas Hi Stuart, can you please check the updated code, I've added a `getPriority` method, so far only to `HttpAuthenticationMechanism` - it can be pushed down to `AuthenticationMechanism` later on, for now the doc note only makes sense for the HTTP ones, so I'm not sure it is worth going for the `quarkus-security` updates, thanks